lms/static/scripts/frontend_apps/utils/api.ts

```diff
@@ -279,35 +301,58 @@ export function usePolledAPIFetch<T = unknown>({
     // When the actual request is loading, transition to loading, in case
     // a new request was triggered by a path/fetch key change
     if (result.isLoading) {
+      // If the fetch is loading but not due to a refresh, it means we are
+      // fetching from a new source. In that case reset the refresh count and
+      // prev data
+      if (!refreshing.current) {
+        setPrevData(null);
+        refreshCount.current = 0;
+      }
+
       setIsLoading(true);
       return () => {};
     }
 
-    // Transition to not-loading only once we no longer have to refresh
-    if (!shouldRefresh(result)) {
+    refreshing.current = false;
+    const maxRefreshesReached = refreshCount.current >= maxRefreshes;
+    setCanceled(maxRefreshesReached);
+
+    // Stop polling once max refreshes is reached or we no longer have to
+    // refresh
+    if (maxRefreshesReached || !shouldRefresh(result)) {
       setIsLoading(false);
-      // Reset prev data once we finish refreshing, so that we don't
-      // accidentally leak data from a different request on subsequent ones
-      setPrevData(null);
       return () => {};
     }
 
     // Once we finish loading, schedule a retry if the request should be
     // refreshed
+    refreshCount.current++;
     timeout.current = _setTimeout(() => {
+      // Track the fact that we are refreshing, in order to know if the fetch is
+      // loading due to a refresh or due to an unrelated change in `path` and/or
+      // `params`
+      refreshing.current = true;
```

Comment:
There is a race condition here if the fetch parameters change in between the call to `retry` and the re-render that it schedules. I think the risk is low though since that re-render happens "immediately" (via something like `queueMicrotask`).

A possible solution would be to expose the fetch key from the `useFetch` result and compare that to the key associated with the refresh.